# Exercice: 

Le premier ensemble de données contient des données quotidiennes sur la météo et l’énergie générée par les panneaux solaires à un endroit particulier (fichier: `power_and_weather.csv`). Voici les variables qu’il contient :

`temperature_min:` la température minimale en ºC.  
`temperature_max:` la température maximale en ºC.  
`humidity_min:` l’humidité minimale  
`humidity_max:` l’humidité maximale  
`rain_vol:` le volume de pluie en mm.  
`kwh:` la quantité d’énergie produite.  

In [1]:
import pandas as pd

**Question 1:** importer les données et afficher les 7 premiers lignes

In [77]:
df = pd.read_csv('power_and_weather.csv')
df

date  rain_vol  temperature_min  temperature_max  humidity_min  \
0    2021-01-01       0.0             21.0             31.2          42.0   
1    2021-01-02       6.8             19.6             29.2          56.0   
2    2021-01-03      14.0             18.2             26.5          70.0   
3    2021-01-04      48.6             18.2             24.0          76.0   
4    2021-01-05       0.0             20.1             28.5          60.0   
..          ...       ...              ...              ...           ...   
345  2021-12-24      28.0             18.6             23.1          70.0   
346  2021-12-26       0.0             18.6             29.0          47.0   
347  2021-12-27       0.0             20.2             29.6          51.0   
348  2021-12-28       0.8             19.3             28.2          62.0   
349  2021-12-31       0.2             20.7             25.8          73.0   

     humidity_max   kwh  day month    quarter  
0            82.0  45.1    1   Jan  Quarter 1  
1            93.0  33.0    2   Jan  Quarter 1  
2            96.0  28.8    3   Jan  Quarter 1  
3            96.0  23.3    4   Jan  Quarter 1  
4            95.0  31.5    5   Jan  Quarter 1  
..            ...   ...  ...   ...        ...  
345          98.0  17.6   24   Dec  Quarter 4  
346          91.0  46.6   26   Dec  Quarter 4  
347          92.0  41.3   27   Dec  Quarter 4  
348          96.0  37.9   28   Dec  Quarter 4  
349          93.0  20.4   31   Dec  Quarter 4  

[350 rows x 10 columns]

**Question 2:** Afficher le type de données de chaque colonne

In [15]:
df.dtypes

date                object
rain_vol           float64
temperature_min    float64
temperature_max    float64
humidity_min       float64
humidity_max       float64
kwh                float64
day                  int64
month               object
quarter             object
dtype: object

__Question 3:__ Que represente le type de données `object`? Quels sont les type de données de base en Pandas?

En Pandas, le type `object` est généralement utilisé pour stocker des données de type texte (chaînes de caractères), mais il peut aussi contenir d’autres types de données Python comme des listes ou des dictionnaires. Lorsque vous voyez un dtype `object` dans un DataFrame Pandas, cela signifie que la colonne contient des données qui sont traitées comme des objets Python génériques.

Voici quelques points clés à retenir sur le type object en Pandas :

`Polyvalence :` Une colonne de type object peut mélanger des données textuelles et non textuelles.  
`Compatibilité :` Le type object assure une compatibilité avec d’autres bibliothèques Python.  
`Efficacité :` Bien que le type object soit plus flexible, il peut être moins efficace en termes de stockage et de vitesse par rapport aux types de données spécifiques comme `int` ou `string`.

[Type de données texte en Pandas](https://pandas.pydata.org/pandas-docs/stable/user_guide/text.html#text-types)  
[Type de données basique en Pandas](https://pandas.pydata.org/pandas-docs/stable/user_guide/basics.html#basics-dtypes)

__Question 4:__ Convertir le type de données de la colonne date en `datetime64`

In [19]:
df.date = pd.to_datetime(df.date)

In [22]:
df.dtypes

date               datetime64[ns]
rain_vol                  float64
temperature_min           float64
temperature_max           float64
humidity_min              float64
humidity_max              float64
kwh                       float64
day                         int64
month                      object
quarter                    object
dtype: object

__Question 5:__ Est il possible de charger le fichier, en spécifiant directement le type `datetime` pour la colonne `date`

In [49]:
df = pd.read_csv('power_and_weather.csv', parse_dates= ['date'])
df.dtypes

date               datetime64[ns]
rain_vol                  float64
temperature_min           float64
temperature_max           float64
humidity_min              float64
humidity_max              float64
kwh                       float64
day                         int64
month                      object
quarter                    object
dtype: object

__Question 6:__ Comment est il possible d'importer le ficher en spécifiant le type de données des colonnes lors du chargement?

In [51]:
# Define the data types for each column
dtype_dict = {
    'month': 'string',
    'quarter': str,
}
df = pd.read_csv('power_and_weather.csv', parse_dates= ['date'], dtype = dtype_dict)
df.dtypes

date               datetime64[ns]
rain_vol                  float64
temperature_min           float64
temperature_max           float64
humidity_min              float64
humidity_max              float64
kwh                       float64
day                         int64
month              string[python]
quarter                    object
dtype: object

__Question 7:__ Peut on utiliser la colonne date comme index. si oui, comment?

In [78]:
df = pd.read_csv('power_and_weather.csv', index_col='date', dtype = dtype_dict)
df

rain_vol  temperature_min  temperature_max  humidity_min  \
date                                                                   
2021-01-01       0.0             21.0             31.2          42.0   
2021-01-02       6.8             19.6             29.2          56.0   
2021-01-03      14.0             18.2             26.5          70.0   
2021-01-04      48.6             18.2             24.0          76.0   
2021-01-05       0.0             20.1             28.5          60.0   
...              ...              ...              ...           ...   
2021-12-24      28.0             18.6             23.1          70.0   
2021-12-26       0.0             18.6             29.0          47.0   
2021-12-27       0.0             20.2             29.6          51.0   
2021-12-28       0.8             19.3             28.2          62.0   
2021-12-31       0.2             20.7             25.8          73.0   

            humidity_max   kwh  day month    quarter  
date                                                  
2021-01-01          82.0  45.1    1   Jan  Quarter 1  
2021-01-02          93.0  33.0    2   Jan  Quarter 1  
2021-01-03          96.0  28.8    3   Jan  Quarter 1  
2021-01-04          96.0  23.3    4   Jan  Quarter 1  
2021-01-05          95.0  31.5    5   Jan  Quarter 1  
...                  ...   ...  ...   ...        ...  
2021-12-24          98.0  17.6   24   Dec  Quarter 4  
2021-12-26          91.0  46.6   26   Dec  Quarter 4  
2021-12-27          92.0  41.3   27   Dec  Quarter 4  
2021-12-28          96.0  37.9   28   Dec  Quarter 4  
2021-12-31          93.0  20.4   31   Dec  Quarter 4  

[350 rows x 9 columns]

__Question 8:__ calculer les ocurrences de elements d'indice date

In [90]:
pd.set_option('display.max_rows', 60)
df.groupby(df.index).size()

date
2021-01-01    1
2021-01-02    1
2021-01-03    1
2021-01-04    1
2021-01-05    1
             ..
2021-12-24    1
2021-12-26    1
2021-12-27    1
2021-12-28    1
2021-12-31    1
Length: 344, dtype: int64

In [81]:
df.index.unique()

Index(['2021-01-01', '2021-01-02', '2021-01-03', '2021-01-04', '2021-01-05',
       '2021-01-06', '2021-01-07', '2021-01-08', '2021-01-09', '2021-01-10',
       ...
       '2021-12-19', '2021-12-20', '2021-12-21', '2021-12-22', '2021-12-23',
       '2021-12-24', '2021-12-26', '2021-12-27', '2021-12-28', '2021-12-31'],
      dtype='object', name='date', length=344)

__Question 9:__ identifier les indices en double, et afficher la table resultante

In [95]:
def filter_func(x):
    return len(x) > 1

In [96]:
df.groupby(df.index).filter(filter_func)

rain_vol  temperature_min  temperature_max  humidity_min  \
date                                                                   
2021-03-26       0.0             20.0             30.4          45.0   
2021-03-26       0.0             20.0             30.4          45.0   
2021-04-07       7.4             18.8             25.9          61.0   
2021-04-07       7.4             18.8             25.9          61.0   
2021-05-11       0.0             14.2             24.7          46.0   
2021-05-11       0.0             14.2             24.7          46.0   
2021-05-21       0.0             16.2             28.3          44.0   
2021-05-21       0.0             16.2             28.3          44.0   
2021-11-15       0.0             16.2             23.8          55.0   
2021-11-15       0.0             16.2             23.8          55.0   
2021-11-15       0.0             16.2             23.8          55.0   

            humidity_max   kwh  day month    quarter  
date                                                  
2021-03-26          94.0  35.0   26   Mar  Quarter 1  
2021-03-26          94.0  35.0   26   Mar  Quarter 1  
2021-04-07          96.0  23.7    7   Apr  Quarter 2  
2021-04-07          96.0  23.7    7   Apr  Quarter 2  
2021-05-11          94.0  26.7   11   May  Quarter 2  
2021-05-11          94.0  26.7   11   May  Quarter 2  
2021-05-21          96.0  27.0   21   May  Quarter 2  
2021-05-21          96.0  27.0   21   May  Quarter 2  
2021-11-15          91.0  26.5   15   Nov  Quarter 4  
2021-11-15          91.0  26.5   15   Nov  Quarter 4  
2021-11-15          91.0  26.5   15   Nov  Quarter 4

In [98]:
# autre facon

result = df.groupby(df.index).filter(lambda x: len(x) > 1)
result

rain_vol  temperature_min  temperature_max  humidity_min  \
date                                                                   
2021-03-26       0.0             20.0             30.4          45.0   
2021-03-26       0.0             20.0             30.4          45.0   
2021-04-07       7.4             18.8             25.9          61.0   
2021-04-07       7.4             18.8             25.9          61.0   
2021-05-11       0.0             14.2             24.7          46.0   
2021-05-11       0.0             14.2             24.7          46.0   
2021-05-21       0.0             16.2             28.3          44.0   
2021-05-21       0.0             16.2             28.3          44.0   
2021-11-15       0.0             16.2             23.8          55.0   
2021-11-15       0.0             16.2             23.8          55.0   
2021-11-15       0.0             16.2             23.8          55.0   

            humidity_max   kwh  day month    quarter  
date                                                  
2021-03-26          94.0  35.0   26   Mar  Quarter 1  
2021-03-26          94.0  35.0   26   Mar  Quarter 1  
2021-04-07          96.0  23.7    7   Apr  Quarter 2  
2021-04-07          96.0  23.7    7   Apr  Quarter 2  
2021-05-11          94.0  26.7   11   May  Quarter 2  
2021-05-11          94.0  26.7   11   May  Quarter 2  
2021-05-21          96.0  27.0   21   May  Quarter 2  
2021-05-21          96.0  27.0   21   May  Quarter 2  
2021-11-15          91.0  26.5   15   Nov  Quarter 4  
2021-11-15          91.0  26.5   15   Nov  Quarter 4  
2021-11-15          91.0  26.5   15   Nov  Quarter 4

__Question 10:__ enlever les valeurs en double (toutes colonnes ensembles)

In [100]:
df.drop_duplicates()

rain_vol  temperature_min  temperature_max  humidity_min  \
date                                                                   
2021-01-01       0.0             21.0             31.2          42.0   
2021-01-02       6.8             19.6             29.2          56.0   
2021-01-03      14.0             18.2             26.5          70.0   
2021-01-04      48.6             18.2             24.0          76.0   
2021-01-05       0.0             20.1             28.5          60.0   
...              ...              ...              ...           ...   
2021-12-24      28.0             18.6             23.1          70.0   
2021-12-26       0.0             18.6             29.0          47.0   
2021-12-27       0.0             20.2             29.6          51.0   
2021-12-28       0.8             19.3             28.2          62.0   
2021-12-31       0.2             20.7             25.8          73.0   

            humidity_max   kwh  day month    quarter  
date                                                  
2021-01-01          82.0  45.1    1   Jan  Quarter 1  
2021-01-02          93.0  33.0    2   Jan  Quarter 1  
2021-01-03          96.0  28.8    3   Jan  Quarter 1  
2021-01-04          96.0  23.3    4   Jan  Quarter 1  
2021-01-05          95.0  31.5    5   Jan  Quarter 1  
...                  ...   ...  ...   ...        ...  
2021-12-24          98.0  17.6   24   Dec  Quarter 4  
2021-12-26          91.0  46.6   26   Dec  Quarter 4  
2021-12-27          92.0  41.3   27   Dec  Quarter 4  
2021-12-28          96.0  37.9   28   Dec  Quarter 4  
2021-12-31          93.0  20.4   31   Dec  Quarter 4  

[344 rows x 9 columns]

__Question 11:__ modifier la valeur Quarter 1 en Q1, Quarter 2 en Q2, etc...

In [101]:
df.quarter = df.quarter.str.replace('Quarter ', 'Q')

__Question 13:__ arrondir les valeur de la colonne `rain_vol` a seulement un chiffre après la virgule

In [105]:
df.rain_vol = round(df.rain_vol,1)

__Question 14:__ enregistrer le dataframe resultant dans un fichier nommeé `power_and_weather_modified`

In [106]:
df.to_csv('power_and_weather_modified.csv')

__Question 15:__ Convertir le type de donnée de la colonne `quarter` en type `str` puis `string`

In [36]:
df.quarter = df['quarter'].astype('string')

In [37]:
df.dtypes

date               datetime64[ns]
rain_vol                  float64
temperature_min           float64
temperature_max           float64
humidity_min              float64
humidity_max              float64
kwh                       float64
day                         int64
month                      object
quarter            string[python]
dtype: object

**Question 7:** Afficher la moyenne de `temperature_max` arrondi à deux chiffres après la virgule

In [13]:
round(df.temperature_max.mean(),2)

27.45

**Question 8:**  Afficher la moyenne de `temperature_max` arrondi à deux chiffres après la virgule pour le premier quart

In [56]:
round(df.temperature_max.loc[df.quarter == 'Quarter 1'].mean(),2)

29.27

In [9]:
df.loc[(df.rain_vol > 20) & (df.month == 'Jan')]

date  rain_vol  temperature_min  temperature_max  humidity_min  \
3  2021-01-04      48.6             18.2             24.0          76.0   

   humidity_max   kwh  day month    quarter  
3          96.0  23.3    4   Jan  Quarter 1

Le tableau ci-dessous contient la température maximale moyenne et la production d’électricité par trimestre de l’année :
Quartier|	temperature_max|	Kwh
--|--|--
1|	29.27|	32.88
2|	26.45|	25.19
3|	27.11|	24.36
4|	26.92|	26.77

In [70]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   date             350 non-null    datetime64[ns]
 1   rain_vol         350 non-null    float64       
 2   temperature_min  350 non-null    float64       
 3   temperature_max  350 non-null    float64       
 4   humidity_min     350 non-null    float64       
 5   humidity_max     350 non-null    float64       
 6   kwh              350 non-null    float64       
 7   day              350 non-null    int64         
 8   month            350 non-null    string        
 9   quarter          350 non-null    object        
dtypes: datetime64[ns](1), float64(6), int64(1), object(1), string(1)
memory usage: 27.5+ KB


__Question 1:__ Reproduire ce tableau

In [58]:
df.groupby('quarter')['temperature_max','kwh'].mean()

ValueError: Cannot subset columns with a tuple with more than one element. Use a list instead.

In [71]:
df.groupby('quarter')[['temperature_max','temperature_min']].mean()

temperature_max  temperature_min
quarter                                    
Quarter 1        29.271429        20.114286
Quarter 2        26.450000        16.763043
Quarter 3        27.112048        16.878313
Quarter 4        26.915476        19.540476

Source: [link](https://www.dataquest.io/blog/data-visualization-techniques/)

In [ ]:
https://www.dataquest.io/blog/data-visualization-techniques/

In [ ]:
# Rename the column in place
df.rename(columns={'old_column_name': 'new_column_name'}, inplace=True)